sklearn/tree/_splitter.pyx

```diff
@@ -86,10 +86,20 @@ cdef class Splitter:
         self.random_state = random_state
 
     def __getstate__(self):
-        return {}
+        return {
+            "criterion": self.criterion,
+            "max_features": self.max_features,
+            "min_samples_leaf": self.min_samples_leaf,
+            "min_weight_leaf": self.min_weight_leaf,
+            "random_state": self.random_state,
+        }
```

Comment:
`Splitter` has a lot of other (scalar) attribute:
https://github.com/scikit-learn/scikit-learn/blob/9017c701833114a75903f580dd0772e1d8d7d125/sklearn/tree/_splitter.pxd#L31-L58

Do we want to integrate those in?

Moreover, should we also extend `__getstate__` and `__setstate__` for the following sub- Cython extension types, as they have extra attributes?

 - `BaseDenseSplitter`:
  https://github.com/scikit-learn/scikit-learn/blob/9017c701833114a75903f580dd0772e1d8d7d125/sklearn/tree/_splitter.pyx#L217-L220
 - `BaseSparseSplitter`:
  https://github.com/scikit-learn/scikit-learn/blob/9017c701833114a75903f580dd0772e1d8d7d125/sklearn/tree/_splitter.pyx#L751-L760

Reply:
I do not think those attributes need to be a part of the state because they are not need during object creation in `__init__`. These extra attributes are populated when the `init` method is called:

https://github.com/scikit-learn/scikit-learn/blob/9017c701833114a75903f580dd0772e1d8d7d125/sklearn/tree/_splitter.pyx#L94-L99

Functionally, I think serializing these attributes is not useful because they will always be overwritten when `init` is called. Currently, the API contract is to always call `init` and then call `node_*` methods.

(I'm guessing this is why they were not including in the `reduce` function before)

Reply:
OK, this might be made more natural potentially, but would necessitate a major refactoring (which is way out of the scope of the PR) if we want to proceed with this rework.